vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1358 @@
+                else
+                    cout<<"at iter "<<b.iter<<", avg hamming loss: "<<b.hamming_loss*1./b.iter<<endl;
+            }
+
+            clock_t begin = clock();
+        
+            if (b.current_pass < 1){ //in the first pass, we need to store the memory:
+                example* new_ec = &calloc_or_throw<example>();
+                copy_example_data(new_ec, &ec, b.oas);
+                remove_repeat_features_in_ec(*new_ec); ////sort unique.
+                b.examples.push_back(new_ec);   
+                if(b.online == true)
+                    insert_example_hal(b, base, b.examples.size() - 1,*b.examples[b.examples.size()-1]); //query and learn
+                
+                insert_example(b, base, b.examples.size() - 1); //unsupervised learning. 
+                for (uint32_t i = 0; i < b.dream_repeats; i++)
+                    experience_replay(b, base);
+            }
+            else{ //starting from the current pass, we just learn using reinforcement signal, no insertion needed:
+                size_t ec_id = (b.iter)%b.examples.size();
+                insert_example_hal(b, base, ec_id, *b.examples[ec_id]); //no insertion will happen in this call
+		        for (uint32_t i = 0; i < b.dream_repeats; i++)
+		            experience_replay(b, base);
+            }
+            b.construct_time += double(clock() - begin)/CLOCKS_PER_SEC;   
+        }
+        else if (b.test_mode == true){
+            b.iter++;
+            if (b.iter % 5000 == 0){
+                if(b.oas == false)
+                    cout<<"at iter "<<b.iter<<", pred error: "<<b.num_mistakes*1./b.iter<<endl;
+                else
+                    cout<<"at iter "<<b.iter<<", avg hamming loss: "<<b.hamming_loss*1./b.iter<<endl;
+            }
+            clock_t begin = clock();
+            predict(b, base, ec);
+            b.test_time += double(clock() - begin)/CLOCKS_PER_SEC;
+        }
+
+    } 
+
+
+    void end_pass(memory_tree& b){
+        b.current_pass ++;
+        cout<<"######### Current Pass: "<<b.current_pass<<", with number of memories strored so far: "<<b.examples.size()<<endl;
+    }
+
+    void finish(memory_tree& b)
+    {
+        for (size_t i = 0; i < b.nodes.size(); ++i)
+            b.nodes[i].examples_index.delete_v();
+        b.nodes.delete_v();
+        for (size_t i = 0; i < b.examples.size(); i++)
+            free_example(b.examples[i]);
+        b.examples.delete_v();
+        cout<<b.max_nodes<<endl;
+        cout<<b.construct_time<<" "<<b.test_time<<endl;
+    }
+
+    ///////////////////Save & Load//////////////////////////////////////
+    ////////////////////////////////////////////////////////////////////
+    #define writeit(what,str)                               \
```

Comment:
Can we stick these in io_buf.h and remove from here and recall_tree.h?

Reply:
I sticked these to io_buf.h, but I didn't touch recall_tree files